src/index.ts

```diff
@@ -34,44 +34,41 @@ export interface IterableEx<T> extends Iterable<T> {
     readonly toArray: () => readonly T[]
     readonly reverse: () => readonly T[]
     readonly isEmpty: () => boolean
-    readonly uniq: (key?: (v: T) => unknown) => IterableEx<T>
+    readonly uniq: (key?: (v: T) => unknown) => IterableEx<T>,
 }
 
-// tslint:disable-next-line:no-class
-class IterableImpl<T> implements IterableEx<T> {
-    public readonly [Symbol.iterator]: () => Iterator<T>
-    constructor(createIterator: () => Iterator<T>) {
-        // tslint:disable-next-line:no-expression-statement
-        this[Symbol.iterator] = createIterator
+export const iterable = <T>(createIterator: () => Iterator<T>): IterableEx<T> => {
+    const property = <P extends readonly unknown[], R>(f: (self: Iterable<T>, ...p: P) => R) =>
+        (...p: P) => f(result, ...p)
```

Comment:
can you rename `result` to something slightly more descriptive? also does the `extends unkown[]` really bring anything here?

Reply:
yes, TypeScript requires that `P` is an array

Reply:
oh I see

Reply:
The name `result` is a convention, similar to `this`. The `result` const is a result of the function.